### PR TITLE
5022 subject area migration improvement

### LINKED ISSUE #5022: add a select for "subject areas"
**Describe the feature you'd like and what it will do**
Currently, subject areas is free text.
Should be a typeahead -- noninvasive.
@martinseul suggested that we should populate a list similar to
https://has.arizona.edu/research-focus-areas

Related to https://github.com/hydroshare/hydroshare/issues/4733

### PR DESCRIPTION
**Pull request by devincowan**
Resolves #5022 by improving the migration of Subject Areas in the database.
Original migration in PR https://github.com/hydroshare/hydroshare/pull/5036 failed to account for odd strings that might be entered as Subject Areas.
This PR doesn't handle all possible cases, but I think it should do fine with all of the Subject Areas currently in our prod db.

### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [ ] Automated Testing

- [ ] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [x] Peer Code review and approval

### Positive Test Case
1. Checkout 2.5.2 or similar previous release
2. Create a user profile with difficult to parse Subject Areas, something ike: "{ this , , will ,, fail{ to parse []}}}"
3. Checkout this branch
4. Run migrations
5. See that the migration succeeds and subject areas are parsed reasonably

### CODE REVIEW

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-5067)

**Comment by devincowan** [comment]
@sblack-usu here is a deploy with the migration:
https://new-hs-ci.hydroshare.org/view/02-Deployment/job/Beta%20deployment/job/beta-deploy-by-branch/523/console

File: theme/migrations/0023_alter_userprofile_subject_areas.py [hunk 1]
```diff
@@ -21,33 +34,50 @@ def migrate_csv_subject_areas(apps, schema_editor):
         print(f'Searching user #{profile.pk} which has subject areas: {profile.subject_areas}')
         new_subj_areas = []
         for subject in old_subject_areas:
+            if subject == '':
+                # There is a trailing comma that we need to remove
+                continue
+            stripped_subject = strip_for_dict(subject)
+            if stripped_subject == '':
+                # The subject contained only invalid chars
+                print(f"- Unmatched subject area '{subject}' contains invalid chars that will be stripped")
```
Comment (sblack-usu) [inline theme/migrations/0023_alter_userprofile_subject_areas.py:43]: Is this a true statement?  There is a continue so the

Comment (sblack-usu) [inline theme/migrations/0023_alter_userprofile_subject_areas.py:43]: nit: this case removes the subject entirely

**Review by sblack-usu — APPROVED** [review]
This looks good.  Is there validation or formatting rules applied to new subjects that align with these?

**Comment by devincowan** [comment]
The tagsinput:
https://github.com/hydroshare/hydroshare/blob/master/theme/static/js/profile.js#L465
strips whitespace but other characters I tested are allowed.

**Pull request merged**